src/main/java/com/hazelcast/gcp/GcpDiscoveryStrategyFactory.java

```diff
@@ -50,4 +59,69 @@ public DiscoveryStrategy newDiscoveryStrategy(DiscoveryNode discoveryNode, ILogg
         }
         return result;
     }
+
+    /**
+     * Checks if Hazelcast is running on GCP.
+     * <p>
+     * To check if Hazelcast is running on GCP, we first check whether the internal DNS is configured for "google.internal" in
+     * either "/etc/resolv.conf" or "/etc/hosts". Such an approach is not officially documented but seems like a good
+     * enough heuristic to detect a GCP Compute VM Instance. Since it's not the official method, we still need to make
+     * an API call to "metadata.google.internal" which will resolve to a local, non-routable address http://169.254.169.254/.
+     * Finally, we check if there is a service account attached for this instance because without a service account Hazelcast
+     * GCP discovery will not work.
+     *
+     * @return true if running on GCP Instance which has a service account attached
+     * @see https://cloud.google.com/compute/docs/instances/managing-instances#dmi
+     */
+    @Override
+    public boolean isAutoDetectionApplicable() {
+        return googleInternalDnsConfigured() && metadataFlavorGoogle() && serviceAccountAttached();
+    }
+
+    private static boolean googleInternalDnsConfigured() {
+        return readFileContents("/etc/resolv.conf").contains("google.internal")
+                || readFileContents("/etc/hosts").contains("google.internal");
+    }
+
+    static String readFileContents(String fileName) {
+        InputStream is = null;
+        try {
+            File file = new File(fileName);
+            byte[] data = new byte[(int) file.length()];
+            is = new FileInputStream(file);
+            is.read(data);
+            return new String(data, StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            throw new RuntimeException("Could not get " + fileName, e);
+        } finally {
+            IOUtil.closeResource(is);
+        }
+    }
+
+    private static boolean metadataFlavorGoogle() {
+        return isEndpointAvailable("metadata.google.internal");
+    }
+
+    private static boolean serviceAccountAttached() {
+        try {
+            return isEndpointAvailable("metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/");
```

Comment:
I think this part is not correct. What is someone does not use `default` service account? Or if someone does not have persmission to check the service accounts?

Reply:
If they have auto-configuration enabled, I assume they don't provide private-key-path. In that case only way to get token for compute api is through a service account. If they don't have any service accounts then GCP discovery won't work.

I don't understand what you mean by "if they don't have permissions to check service accounts"? As far as I know if they are on a GCP VM they can use instance metadata api without need for a permission, because it is only provided for that instance.

For the default service account, in "GcpMetadataApi" class we use default service account to get the token, so if they don't have a default one, discovery api wouldn't work nevertheless.

I am not confident about my findings though I am not familiar with gcp.

Reply:
If you define a different service account while creating a GCP VM, will it still work?
![image](https://user-images.githubusercontent.com/2834997/89423721-2872b280-d737-11ea-81c1-c273f1fd1b67.png)

Reply:
Other than rafal? If the service account has permissions for reading project and region data, then I think it would work.

Reply:
If you checked that it works with other service account, then it's fine.

Reply:
I will check it again.

Reply:
It works correctly, I tried it with two different service accounts